rclcpp/test/rclcpp/test_rate.cpp

```diff
@@ -21,12 +21,24 @@
 
 #include "../utils/rclcpp_gtest_macros.hpp"
 
+class TestRate : public ::testing::Test
+{
+public:
+  void SetUp()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void TearDown()
+  {
+    rclcpp::shutdown();
+  }
+};
+
 /*
    Basic tests for the Rate, WallRate and ROSRate classes.
  */
-TEST(TestRate, rate_basics) {
-  rclcpp::init(0, nullptr);
-
+TEST_F(TestRate, rate_basics) {
   auto period = std::chrono::milliseconds(1000);
   auto offset = std::chrono::milliseconds(500);
```

Comment:
> It seems like this is performance problem: when while sleeping 500ms (with rate period 1000ms), we are actually sleeping more than 1000ms.

according to @AlexeyMerzlyakov, these are too racy for windows platform. probably we should extend this like,

```suggestion
  auto period = std::chrono::milliseconds(2000);
  auto offset = std::chrono::milliseconds(1000);
```

any thoughts?

Reply:
After some analysis, I've discovered that the problem appeared due to being late inside `rclcpp::sleep_for()` waiting cycle: condition variable [sleeps](https://github.com/ros2/rclcpp/blob/7f7ffcf6ed1143b4d4a9e0cf248e190731b2e5f8/rclcpp/src/rclcpp/context.cpp#L499) for a required period of time twice, which determines the multiple increase of sleeping time (from 500ms -> to 1000ms).

So, I'd like to change the `offset` value only, to be equal to `30%` from `period` time in the `rate_basics`, `wall_rate_basics` and `generic_rate` testcases. This will allow even if `rclcpp::sleep_for()` is missed its time, it will sleep only 60% of period, and we won't miss rate timings.

Local testing shown promising failure rate: only 1 failure of 100 runs in a row. This failure rate was related to measured delta after `GenericRate::sleep()` (which uses `rclcpp::sleep_for()` in its sleep implementation) was not in between `(period-epsilon ... 1.5*period)` range. This to be the same `rclcpp::sleep_for()` problem, but to deal with it, we need to change its main cycle mechanism. Anyway, the failure rate after chaning `offset` to be 30% from `period` seems very promising to apply in this test.

Reply:
@AlexeyMerzlyakov thanks for the analysis, addressed in https://github.com/ros2/rclcpp/pull/2329/commits/1097ba03c5c372bca9483dc66388adb7585a12e3.

btw,

> rclcpp::sleep_for() waiting cycle: condition variable [sleeps](https://github.com/ros2/rclcpp/blob/7f7ffcf6ed1143b4d4a9e0cf248e190731b2e5f8/rclcpp/src/rclcpp/context.cpp#L499) for a required period of time twice, which determines the multiple increase of sleeping time (from 500ms -> to 1000ms).

is this because of system tick resolution problem? i am not sure why this would happen.

Reply:
see https://github.com/ros2/rclcpp/pull/2329/commits/bff0d84978f6fbbad0cde62662db46e218bd5ad5

and reverted https://github.com/ros2/rclcpp/pull/2329/commits/8dee8b3bd7358a09804576d26341b6c3f9444ce9

Reply:
Great! The proposed in https://github.com/ros2/rclcpp/commit/bff0d84978f6fbbad0cde62662db46e218bd5ad5 solution is used to sleep with remained amount of time on each iteration instead of fixed. This seems to be correct: even if we missed first cycle, the program will sleep-up remaining period of time.

Run-time verification (on `test_rate`) on Windows confirms that the cycle iteration sequence is being converged for `Context::sleep_for()`: maximum number of iterations for waiting cycle was `2` while average number of iterations during 50 runs - is `1.03`. On the Linux there were no cases with more than `1` cycle iteration were detected for 50 test runs.